### PR TITLE
Public Links

### PR DESCRIPTION
**Pull request by refs**
Provide `reva-storage-public-link` command and its default UNIX configuration.

### CODE REVIEW

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/command/storagepubliclink.go  8
- pkg/flagset/storagepubliclink.go  1
         

Clones added
============
- pkg/command/storagepubliclink.go  9
- pkg/command/users.go  1
- pkg/command/authbasic.go  1
- pkg/command/storageocdata.go  1
- pkg/flagset/storagepubliclink.go  18
- pkg/command/storagehomedata.go  1
- pkg/command/sharing.go  1
- pkg/command/storageoc.go  1
- pkg/command/storagehome.go  1
- pkg/command/authbearer.go  1
- pkg/command/frontend.go  1
- pkg/command/gateway.go  1
- pkg/command/storageroot.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

File: pkg/command/gateway.go [hunk 1]
```diff
@@ -126,16 +128,18 @@ func Gateway(cfg *config.Config) *cli.Command {
...
+											cfg.Reva.StorageEOS.MountID:            cfg.Reva.StorageEOS.URL,
+											cfg.Reva.StorageOC.MountPath:           cfg.Reva.StorageOC.URL,
+											cfg.Reva.StorageOC.MountID:             cfg.Reva.StorageOC.URL,
+											cfg.Reva.StorageS3.MountPath:           cfg.Reva.StorageS3.URL,
+											cfg.Reva.StorageS3.MountID:             cfg.Reva.StorageS3.URL,
+											cfg.Reva.StorageWND.MountPath:          cfg.Reva.StorageWND.URL,
+											cfg.Reva.StorageWND.MountID:            cfg.Reva.StorageWND.URL,
+											cfg.Reva.StorageCustom.MountPath:       cfg.Reva.StorageCustom.URL,
+											cfg.Reva.StorageCustom.MountID:         cfg.Reva.StorageCustom.URL,
+											"/public/":                             "localhost:10054",
```
Comment (refs) [inline pkg/command/gateway.go:141]: this needs to be added to the config / flagset

File: pkg/command/storagepubliclink.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+						"network": cfg.Reva.StoragePublicLink.Network,
+						"address": cfg.Reva.StoragePublicLink.Addr,
+						"interceptors": map[string]interface{}{
+							"log": map[string]interface{}{},
+						},
+						"services": map[string]interface{}{
+							"publicstorageprovider": map[string]interface{}{
+								"mount_path":   "/public/", // localhost:[ocdavport]/remote.php/dav/public-files/{token}/folderA/folderB/files.txt
+								"mount_id":     "e1a73ede-549b-4226-abdf-40e69ca8230d",
+								"gateway_addr": "localhost:9142",
```
Comment (refs) [inline pkg/command/storagepubliclink.go:87]: `gateway_addr` should be read from `config`

File: pkg/command/storagepubliclink.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+							},
+							"authprovider": map[string]interface{}{
+								"auth_manager": "publicshares",
+								"auth_managers": map[string]interface{}{
+									"publicshares": map[string]interface{}{
+										"gateway_addr": "localhost:9142",
```
Comment (refs) [inline pkg/command/storagepubliclink.go:93]: `gateway_addr` should be read from `config`

File: pkg/config/config.go [hunk 4]
```diff
@@ -79,6 +79,15 @@ type StoragePort struct {
 	TempFolder string
 }
 
+// PublicStorage configures a public storage provider
+type PublicStorage struct {
```
Comment (refs) [inline pkg/config/config.go:83]: can be avoided by using the config from the storage provider (?)

File: pkg/flagset/storagepubliclink.go [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+		&cli.StringFlag{
+			Name:        "addr",
+			Value:       "0.0.0.0:10054",
+			Usage:       "Address to bind reva service",
+			EnvVars:     []string{"REVA_STORAGE_PUBLIC_LINK_ADDR"},
+			Destination: &cfg.Reva.StoragePublicLink.Addr,
+		},
+		&cli.StringFlag{
+			Name:        "url",
+			Value:       "localhost:10054",
```
Comment (refs) [inline pkg/flagset/storagepubliclink.go:114]: reserve a port

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/flagset/storagepubliclink.go  1
- pkg/command/storagepubliclink.go  8
         

Clones added
============
- pkg/flagset/storagepubliclink.go  18
- pkg/command/gateway.go  1
- pkg/command/storagepubliclink.go  9
- pkg/command/frontend.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/flagset/storagepubliclink.go  1
- pkg/command/storagepubliclink.go  8
         

Clones added
============
- pkg/command/gateway.go  1
- pkg/command/frontend.go  1
- pkg/flagset/storagepubliclink.go  5
- pkg/command/storagepubliclink.go  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

File: pkg/flagset/sharing.go [hunk 6]
```diff
@@ -136,7 +136,7 @@ func SharingWithConfig(cfg *config.Config) []cli.Flag {
 		},
 		&cli.StringFlag{
 			Name:        "public-driver",
-			Value:       "memory",
+			Value:       "json",
```
Comment (refs) [inline pkg/flagset/sharing.go:139]: the json driver has a hard dependency on its db file. This might fail a default ocis run.

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/command/storagepubliclink.go  8
- pkg/flagset/storagepubliclink.go  1
         

Clones added
============
- pkg/command/frontend.go  2
- pkg/command/storagepubliclink.go  7
- pkg/command/gateway.go  2
- pkg/flagset/storagepubliclink.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/flagset/storagepubliclink.go  1
- pkg/command/storagepubliclink.go  8
         

Clones added
============
- pkg/command/frontend.go  2
- pkg/flagset/storagepubliclink.go  5
- pkg/command/storagepubliclink.go  7
- pkg/command/gateway.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/flagset/storagepubliclink.go  1
- pkg/command/storagepubliclink.go  8
         

Clones added
============
- pkg/command/gateway.go  1
- pkg/command/frontend.go  1
- pkg/flagset/storagepubliclink.go  5
- pkg/command/storagepubliclink.go  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by PVince81** [comment]
- [ ] make link grant path configurable

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/command/storagepubliclink.go  8
- pkg/flagset/storagepubliclink.go  1
         

Clones added
============
- pkg/command/gateway.go  1
- pkg/command/frontend.go  1
- pkg/command/storagepubliclink.go  9
- pkg/flagset/storagepubliclink.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/command/storagepubliclink.go  8
- pkg/flagset/storagepubliclink.go  1
         

Clones added
============
- pkg/command/storagepubliclink.go  9
- pkg/command/frontend.go  1
- pkg/flagset/storagepubliclink.go  4
- pkg/command/gateway.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/flagset/storagepubliclink.go  1
- pkg/command/storagepubliclink.go  8
         

Clones added
============
- pkg/command/frontend.go  1
- pkg/command/gateway.go  1
- pkg/flagset/storagepubliclink.go  4
- pkg/command/storagepubliclink.go  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by refs** [comment]
>     * [ ]  make link grant path configurable

[we got rid of this](https://github.com/cs3org/reva/pull/681/commits/e42564b8936542fbbe909f83088bcb540d0d117f#diff-3025287d6fa62793fd65128d07e74cd4L62)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/flagset/storagepubliclink.go  1
- pkg/command/storagepubliclink.go  8
         

Clones added
============
- pkg/command/storageoc.go  2
- pkg/command/frontend.go  1
- pkg/flagset/storagepubliclink.go  18
- pkg/command/storageocdata.go  2
- pkg/command/storageroot.go  2
- pkg/command/storagepubliclink.go  10
         

Clones removed
==============
+ pkg/command/gateway.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/flagset/storagepubliclink.go  1
- pkg/command/storagepubliclink.go  8
         

Clones added
============
- pkg/command/frontend.go  1
- pkg/command/storageocdata.go  2
- pkg/flagset/storagepubliclink.go  18
- pkg/command/storageroot.go  2
- pkg/command/storagepubliclink.go  10
- pkg/command/storageoc.go  2
         

Clones removed
==============
+ pkg/command/gateway.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5357590&bid=17555040)

**Comment by PVince81** [comment]
seems we still have hard-coded localhost:10054 ? can we sort this out and make it configurable ?

**Comment by PVince81** [comment]
to solve the CI, please rebase on top of origin/master from the owncloud/ocis-reva repo.
it should contain an adjustment in the Drone config related to mimetype setup.

**Comment by PVince81** [comment]
closing in favor of https://github.com/owncloud/ocis-reva/pull/232

**Pull request closed without merging**